### PR TITLE
Feature/modelize component container

### PR DESCRIPTION
**Pull request by Hundraw**
:information_source:  Don't forget to modify the changelog.md before merging this branch.

### CODE REVIEW

File: changelog.md [hunk 1]
```diff
@@ -12,6 +12,8 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 - Add property `definition` in `ComponentAttribute`
 - Add property `definition` in `ComponentLink`
 - Add property `attributeRef` in `ComponentLinkDefinition`
+- Add method `setContainerSize` using `bin-pack` library in `DefaultDrawer`.
+- Add feature that Drawing components in their container and resize it, in  `DefaultDrawer`.
```
Comment (k-y0u) [inline changelog.md:16]: `Drawing` -> `drawing`

File: src/draw/DefaultDrawer.js [hunk 2]
```diff
@@ -9,15 +9,15 @@ import ComponentDrawOption from 'src/models/ComponentDrawOption';
 class DefaultDrawer {
   /**
    * Default constructor
-   * @param {String} [parentId="#root"] - Id of HTML element where we want to draw.
    * @param {Object} [resources=null] - Object that contains resources.
+   * @param {String} [rootId="#root"] - Id of HTML element where we want to draw.
```
Comment (k-y0u) [inline src/draw/DefaultDrawer.js:13]: `rootId="#root"` -> rootId="root"

File: src/draw/DefaultDrawer.js [hunk 3]
```diff
@@ -55,22 +55,39 @@ class DefaultDrawer {
   }
```
Comment (Zorin95670) [inline src/draw/DefaultDrawer.js:55]: typo on `deltay` line 52

File: src/draw/DefaultDrawer.js [hunk 4]
```diff
@@ -261,48 +253,93 @@ class DefaultDrawer {
...
-      .attr('x', (data) => data.drawOption.x)
-      .attr('y', (data) => data.drawOption.y)
-      .attr('cursor', 'grab');
+  __drawModels(components) {
+    const templates = [
+      ...new Set(components.map((component) => component.definition.model)),
+    ];
+
+    templates.forEach((template) => {
+      const elements = this.d3.selectAll(`.component-${template}`);
```
Comment (Zorin95670) [inline src/draw/DefaultDrawer.js:319]: select by rootId too

File: src/draw/DefaultDrawer.js [hunk 5]
```diff
@@ -261,48 +253,93 @@ class DefaultDrawer {
...
-    const bBox = this.d3.select(d3Element).node().getBBox();
+  getComponentSize(componentId) {
+    const bBox = this.d3.select(`#${componentId}`).node().getBBox();
     return { width: bBox.width, height: bBox.height };
   }
 
   /**
-   * Draw component with default template.
-   * @param {Object} d3Element - D3 selected element.
+   * Calls all methods dedicated to each models.
```
Comment (Zorin95670) [inline src/draw/DefaultDrawer.js:309]: model without s

**Review by Zorin95670 — CHANGES_REQUESTED** [review]

File: src/draw/DefaultDrawer.js [hunk 6]
```diff
@@ -166,7 +183,7 @@ class DefaultDrawer {
   }
 
   /**
-   * Unselect current selected element and select element by its id.
+   * Unselects current selected element and select element by its id.
```
Comment (k-y0u) [inline src/draw/DefaultDrawer.js:186]: `and select` -> `and selects`

File: src/draw/DefaultDrawer.js [hunk 7]
```diff
@@ -193,66 +210,41 @@ class DefaultDrawer {
   }
 
   /**
-   * Initialize all draw option of the components.
+   * Initializes all draw option of the components.
```
Comment (k-y0u) [inline src/draw/DefaultDrawer.js:213]: `all draw option` -> `all draw options`

File: src/draw/DefaultDrawer.js [hunk 8]
```diff
@@ -193,66 +210,41 @@ class DefaultDrawer {
...
+   * @param {Object} d3Element - D3 selected elements.
```
Comment (k-y0u) [inline src/draw/DefaultDrawer.js:214]: `D3 selected elements.` -> `D3 selected element.`

File: src/draw/DefaultDrawer.js [hunk 9]
```diff
@@ -193,66 +210,41 @@ class DefaultDrawer {
...
    * @param {Component[]} components - List of Components.
-   * @param {Object} d3Element - D3 selected element.
+   * @param {String} parentId - ID of parent it contains components.
```
Comment (k-y0u) [inline src/draw/DefaultDrawer.js:216]: `ID of parent it contains components.` -> `ID of parent which contains components.`

File: src/draw/DefaultDrawer.js [hunk 10]
```diff
@@ -193,66 +210,41 @@ class DefaultDrawer {
...
    */
-  initializeComponents(components, d3Element) {
-    let componentsToInit = components.filter((component) => {
+  initializeComponents(d3Elements, components, parentId) {
```
Comment (k-y0u) [inline src/draw/DefaultDrawer.js:218]: `d3Elements` not accordingly to the description of the method.

File: src/draw/DefaultDrawer.js [hunk 11]
```diff
@@ -193,66 +210,41 @@ class DefaultDrawer {
...
+      .select('.template')
+      .attr('width', '100%')
+      .attr('height', '100%');
   }
 
   /**
-   * Defines the position of the components by using bin-pack library
-   * @param {Component} components - Components we want to set the position.
+   * Sets the containers sizes
+   * @param {Component[]} components - Array Contains components to init the drawOption.
```
Comment (k-y0u) [inline src/draw/DefaultDrawer.js:243]: `Contains` -> `containing`

File: src/draw/DefaultDrawer.js [hunk 12]
```diff
@@ -261,48 +253,93 @@ class DefaultDrawer {
...
+      component.drawOption.y = position.y + this.margin;
     });
   }
 
   /**
-   * Set size of the component from the reel size of the svg model.
-   * @param {Object} d3Element - HTML element selected contained d3Element.
-   * @param {Component} component - Component object we want to set the size.
+   * Stores in the data the size of the component from the size of the svg model.
+   * @param {Component[]} components - Array of component we want to store the size in datas.
```
Comment (k-y0u) [inline src/draw/DefaultDrawer.js:288]: `size in datas` -> `size in data`

File: src/draw/DefaultDrawer.js [hunk 13]
```diff
@@ -261,48 +253,93 @@ class DefaultDrawer {
...
+      }
+    });
+  }
 
-    d3Element.select('.component-name').text((data) => data.name);
-    d3Element.select('.component-type').text((data) => data.definition.type);
-    d3Element.select('.component-icon')
+  /**
+   * Draws component with default template.
+   * @param {Object} d3Elements - D3 selected element.
```
Comment (k-y0u) [inline src/draw/DefaultDrawer.js:328]: `d3Elements` is plural but `selected element` is singular. Issue ?

File: src/draw/DefaultDrawer.js [hunk 14]
```diff
@@ -261,48 +253,93 @@ class DefaultDrawer {
...
+  drawDefaultModel(d3Elements) {
+    d3Elements.select('.component-name').text((data) => data.name);
+    d3Elements.select('.component-type').text((data) => data.definition.type);
+    d3Elements.select('.component-icon')
       .html((data) => this.resources.icons[data.definition.icon]);
   }
+
+  /**
+   * Draws component container with default template.
+   * @param {Object} d3Elements - D3 selected element.
```
Comment (k-y0u) [inline src/draw/DefaultDrawer.js:339]: Same here.

File: src/draw/DefaultDrawer.js (hunk 4 above)
Comment (Zorin95670) [reply src/draw/DefaultDrawer.js:319]: as we discuss, we add layer in another pr

**Review by k-y0u — CHANGES_REQUESTED** [review]

File: src/draw/DefaultDrawer.js (hunk 13 above)
Comment (Hundraw) [reply src/draw/DefaultDrawer.js:328]: that's how d3 library works
```js
d3Elments = d3
    .selectAll('class of element exist or not yet')
    .data('data to link', (data) => data.id)
    .enter() // Check if new data is put or if the data exists yet (it's like a loop i guess)
    .append('type of html element we want drawing')
    .etc() ...
```

File: src/draw/DefaultDrawer.js (hunk 5 above)
Comment (Hundraw) [reply src/draw/DefaultDrawer.js:309]: Done

File: changelog.md (hunk 1 above)
Comment (Hundraw) [reply changelog.md:16]: Done

File: src/draw/DefaultDrawer.js (hunk 2 above)
Comment (Hundraw) [reply src/draw/DefaultDrawer.js:13]: Done

File: src/draw/DefaultDrawer.js (hunk 6 above)
Comment (Hundraw) [reply src/draw/DefaultDrawer.js:186]: Done

File: src/draw/DefaultDrawer.js (hunk 7 above)
Comment (Hundraw) [reply src/draw/DefaultDrawer.js:213]: Done

File: src/draw/DefaultDrawer.js (hunk 8 above)
Comment (Hundraw) [reply src/draw/DefaultDrawer.js:214]: Done

File: src/draw/DefaultDrawer.js (hunk 9 above)
Comment (Hundraw) [reply src/draw/DefaultDrawer.js:216]: Done

File: src/draw/DefaultDrawer.js (hunk 10 above)
Comment (Hundraw) [reply src/draw/DefaultDrawer.js:218]: Done

File: src/draw/DefaultDrawer.js (hunk 11 above)
Comment (Hundraw) [reply src/draw/DefaultDrawer.js:243]: Done

File: src/draw/DefaultDrawer.js (hunk 12 above)
Comment (Hundraw) [reply src/draw/DefaultDrawer.js:288]: Done

File: src/draw/DefaultDrawer.js (hunk 14 above)
Comment (Hundraw) [reply src/draw/DefaultDrawer.js:339]: Same as the upper conversation

File: src/draw/DefaultDrawer.js (hunk 3 above)
Comment (Hundraw) [reply src/draw/DefaultDrawer.js:55]: Done

File: src/draw/DefaultDrawer.js (hunk 6 above)
Comment (k-y0u) [reply src/draw/DefaultDrawer.js]: Not done correctly

File: src/draw/DefaultDrawer.js (hunk 8 above)
Comment (k-y0u) [reply src/draw/DefaultDrawer.js]: `{Object} d3Elements` shouldn't it be an array ?

File: src/draw/DefaultDrawer.js (hunk 13 above)
Comment (k-y0u) [reply src/draw/DefaultDrawer.js:328]: `{Object} d3Elements - D3 selected elements.` L.214

Comment (Zorin95670) [reply src/draw/DefaultDrawer.js:328]: it's more understanable for the users that will use this methods if d3Elements in singular :)

**Review by Zorin95670 — CHANGES_REQUESTED** [review]

File: src/draw/DefaultDrawer.js (hunk 13 above)
Comment (Hundraw) [reply src/draw/DefaultDrawer.js]: Done

File: src/draw/DefaultDrawer.js (hunk 14 above)
Comment (Hundraw) [reply src/draw/DefaultDrawer.js]: Done

File: src/draw/DefaultDrawer.js (hunk 6 above)
Comment (Hundraw) [reply src/draw/DefaultDrawer.js]: Done correctly

Comment (k-y0u) [reply src/draw/DefaultDrawer.js]: Indeed BUT you forgot to put back your first correction ! ;)

File: src/draw/DefaultDrawer.js (hunk 13 above)
Comment (k-y0u) [reply src/draw/DefaultDrawer.js]: @Zorin95670 , so it should be like this all over the file ?
`{Object} d3Element - D3 selected element.`

Comment (Zorin95670) [reply src/draw/DefaultDrawer.js]: let me check

Comment (Zorin95670) [reply src/draw/DefaultDrawer.js]: plural for this methods:
- initializeComponents

Singular for this methods:
- drawDefaultModel
- drawDefaultContainer

**Review by k-y0u — APPROVED** [review]

**Review by Zorin95670 — APPROVED** [review]

**Pull request merged**